lab2/lab2.md

```diff
@@ -1,30 +1,624 @@
+You only need to implement the Branch instructions. This will entail making appropriate connection between the:
+1. Main Control Unit ( only the branch control signal)
+2. Branch Control Unit
+3. Imm Generator
+4. Register file
+5. Program Counter
+6. Instruction memory
+
+
+## Testing the branch instuctions
+To run the tests, you execute the `SingleCycleBranchTesterLab2` suite as follows:
+
+```
+dinocpu:sbt> testOnly Lab2 / dinocpu.SingleCycleRTypeTesterLab2
+```
+
+This will load some binary applications from `src/test/resources/risc-v`.
+The applications that it is running is specified in the output.
+Below is an example of a test that failed:
+
+```
+[info] SingleCycleBranchTesterLab2:
+[info] Single Cycle CPU
+[info] - should run rtype add1 *** FAILED ***
+...
+```
+
+This ran a **branch** application which used the binary **beq**.
+You can view the RISC-V assembly for this application in `src/test/resources/risc-v/bne.riscv`.
+The list of applications that this suite will run can be found in the `InstTests.scala` file (`src/test/scala/cpu-tests/InstTests.scala`).
+If you want more details on the syntax and how to extend this to other RISC-V binaries, ask on Piazza and we will be happy to expand this section.
+
+If you want to run only a single application from this suite of tests, you can add a parameter to the `test` sbt task.
+You can pass the option `-z` which will execute any tests that match the text given to the parameter.
+You must use `--` between the parameters to the sbt task (e.g., the suite to run) and the parameters for testing.
+For instance, to only run the branch less than test, you would use the following:
+
+```
+sbt> testOnly Lab2 / dinocpu.SingleCycleBranchTesterLab2 -- -z blt
+```
+# Part II: Implement the Control Unit
+
+**The test for this part is `dinocpu.ControlTesterLab2`.**
+
+In this part you will be implementing the main control unit in the CPU design. The control unit which has the instruction as an input , is ued to determine how to set the control lines for the functional units and the the multiplexors.
+
+It takes a single input which is the 7-bit `opcode` and generates 9 control signals as output.
+
+```
+ 
+ branch	:  true if branch or jal and update PC with immediate
+ memread: true if we should read memory
+ toreg	: if writing ALU result, 1 if writing memory data, 2 if writing pc+4
+ add	: true if the ALU should add the results
+ memwrite: write the memory
+ regwrite: write the register file
+ immediate: true if use the immediate value
+ alusrc1: 0 is Read data 1, 1 is zero, 2 is PC
+ jump	: 0 no jump, 2 jump, 3 jump and link register
+```
+
+The following table specifies the `Opcode` format and the control signals to be generated.
+
+|---------|-------------|--------|---------|-------|--------|----------|-----------|----------|---------|------|
+| opcode  |opcode format| branch | memread | toreg |   add  | memwrite | immediate | regwrite | alusrc1 | jump |
```

Comment:
We don't want to give them this table. This is the main part of the assignment. They should come up with all of these values on their own!

Also, we should make sure to erase this from the history of this repo so they can't see it :). Hopefully they aren't following this discussion.

We can give them a couple of examples (e.g., R-types and invalid), but we shouldn't give them everything.